prover/src/provers/final_prover.rs

```diff
@@ -19,35 +20,63 @@ impl FinalProver {
 impl Prover<FinalContext> for FinalProver {
     fn prove(&self, ctx: &FinalContext) -> Result<()> {
         log::info!("start final_stark prove");
+        let mut prove_data_cache = ctx.prove_data_cache.lock().unwrap();
 
         // 1. compress setup
         let rc2 = &ctx.recursive2_circom;
-        let r2 = &ctx.recursive2_stark;
+        let mut r2 = ctx.recursive2_stark.clone();
         let sp = &ctx.final_stark;
         let cc = &ctx.final_circom;
         log::info!("rc2: {:?}", rc2);
         log::info!("r2: {:?}", r2);
         log::info!("sp: {:?}", sp);
         log::info!("cc: {:?}", cc);
-        circom_compiler(
-            rc2.circom_file.clone(),
-            "goldilocks".to_string(),
-            "full".to_string(),
-            rc2.link_directories.clone(),
-            cc.output.clone(),
-            false,
-            false,
-        )?;
+        if prove_data_cache.final_cache.already_cached {
+            r2.r1cs_file
+                .clone_from(&prove_data_cache.final_cache.r1cs_file);
+            r2.pil_file
+                .clone_from(&prove_data_cache.final_cache.pil_file);
+            r2.const_file
+                .clone_from(&prove_data_cache.final_cache.const_file);
+            r2.exec_file
+                .clone_from(&prove_data_cache.final_cache.exec_file);
+        }
```

Comment:
here is a classical if else expression, just use if else.

Reply:
ok, already optimized